psamm/moma.py

```diff
@@ -79,84 +100,332 @@ def __init__(self, model, solver):
 
+    # Returns a generator of all the reactions ids
+    def _all_reactions(self):
+        """Reaturns a generator of all the reactions in the model."""
+        for reaction_id in self._model.reactions:
+            yield reaction_id
+
+    # Solve the linear programming problem
+    # Sense tells the solver to minimize or maximize the result
     def _solve(self, sense=None):
+        """Solves the current problem.
+
+        Args:
+            sense: Tells the solver to either minimize or maximize the
+                objective.
+        """
         # Remove temporary constraints
         while len(self._remove_constr) > 0:
             self._remove_constr.pop().delete()
 
+        # Try to solve the problem
         try:
             return self._prob.solve(sense=sense)
+        # Reset the remove constraints list before the next solve
         finally:
             self._remove_constr = []
 
+    # Solves for the standard FBA of the problem
+    # Objective is the biomass
     def _solve_fba(self, objective):
+        """Solves the problem using FBA."""
         self._prob.set_objective(self._v_wt(objective))
 
+        # Solve and store the result
         result = self._solve(lp.ObjectiveSense.Maximize)
+
+        # If no solution was found an error is raised and told to the user
         if not result:
             raise MOMAError('Unable to solve initial FBA: {}'.format(
                 result.status))
 
-        return result.get_value(self._v_wt(objective))
-
-    def minimize_l1(self, objective):
-        wt_obj = self._solve_fba(objective)
-
+        # Return the result object
+        return result
+
+    # Returns all the flux values for each of the reactions. This is used in
+    # MOMA LP2 and QLP2 to minimize the change in the flux flow following a
+    # gene deletion.
+    def _get_fba_flux(self, objective):
+        """Returns a dictionary of all the solved fluxes for each reaction
+        in the model."""
+        # Get the result of the wild type FBA problem
+        flux_result = self._solve_fba(objective)
+
+        # Creat a dictionary of all the reaction ids and their fluxes.
+        # {'reaction1': 1000, 'reaction2': 2000, ...}
+        fba_fluxes = {}
+
+        # Place all the flux values in a dictionary
+        for key in self._model.reactions:
+            fba_fluxes[key] = flux_result.get_value(self._v_wt(key))
+
+        # Return the dictionary with all of the fluxes associated with the
+        # reaction id
+        # {'reaction1': 1000, 'reaction2': 2000, ...}
+        return fba_fluxes
+
+    def _get_minimal_flux_fba(self, wt_biomass, objective):
+        """Finds the FBA solution that minimizes all the flux values, while
+        still maximizing the biomass. This still might not find a unique
+        solution."""
+        wt_obj = self._v_wt(objective)
+        with self.constraints(wt_biomass == wt_obj):
+            self._prob.set_objective(self._v_wt.sum(self._all_reactions()))
```

Comment:
Unfortunately this will not work since we have reversible reactions with negative fluxes. We are interested in minimizing the sum of absolute fluxes not just the sum of fluxes. See the `minimize_l1()` method in the `fluxanalysis` module (you'll need additional variables to do this).